module/zfs/metaslab.c

```diff
@@ -1459,7 +1661,27 @@ metaslab_load_impl(metaslab_t *msp)
 			    range_tree_remove, msp->ms_allocatable);
 		}
 	}
-	msp->ms_max_size = metaslab_block_maxsize(msp);
+
+	/*
+	 * We expect the ms_allocatable to be a more compact
+	 * representation of the free segments in this metaslab
+	 * compared to the ms_sm (if not just equally compact).
+	 * Therefore, now it is a good time to recalculate the
+	 * metaslab's weight, fragmentation, and ms_max_size
+	 * (the maximum allocatable size).
+	 */
+	uint64_t weight = msp->ms_weight;
+	metaslab_recalculate_weight_and_sort(msp);
+
+	/*
+	 * If there hasn't been any funny business where we just
+	 * upgraded the pool and we converted from a space-based
+	 * to a segment-based one, we expect the ms_allocatable
+	 * to give us a better weight from the ms_sm if not equal.
+	 * [see comment above why].
```

Comment:
from -> than

Rather than referring to the above comment, let's explain here why the "real" weight may be better than expected (due to consolidation of adjacent segments between txg's).  Then I think we can simplify the "above" comment (before `weight` is declared).